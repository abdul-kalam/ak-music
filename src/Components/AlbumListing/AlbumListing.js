import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { AlbumList } from '../AlbumList/AlbumList';
import {
  selectAllMusic,
  selectTittleList,
  setSearchTerm,
  getFilterData,
} from '../../Store/Reducers/musicSlice';
import { Autocomplete } from '../Autocomplete/Autocomplete';
import { Filter } from '../Filter/filter';
import {
  AlbumListContainer,
  AlbumWrapper,
  SearchWrapper,
} from './AlbumListing.style';
import { MessageBanner } from '../../Components/MessageBanner/MessageBanner';

export const AlbumListing = () => {
  const dispatch = useDispatch();
  const albums = useSelector(selectAllMusic);
  const titleList = useSelector(selectTittleList);
  const { searchTerm } = useSelector(getFilterData);

  const onAutoCompleteChange = (searchText) => {
    dispatch(setSearchTerm({ searchTerm: searchText }));
  };

  return (
    <AlbumWrapper>
      <SearchWrapper>
        <Autocomplete
          defaultVal={searchTerm}
          dataList={titleList}
          onSearch={onAutoCompleteChange}
        />
      </SearchWrapper>
      <Filter />
      <AlbumListContainer>
        {albums.length ? (
          <AlbumList showList={albums} />
        ) : (
          <MessageBanner type="info">
            Albums not available for selected criteria. Try something else!
          </MessageBanner>
        )}
      </AlbumListContainer>
    </AlbumWrapper>
  );
};
