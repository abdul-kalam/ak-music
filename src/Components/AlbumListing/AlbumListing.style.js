import styled from 'styled-components';

export const AlbumListContainer = styled.section`
    margin-top: 24px;
    margin-bottom: 24px;
`;
export const AlbumWrapper= styled.section`
    display: flex;
    flex-direction: column;
    justify-content:center;
    align-items: center;
    margin-bottom: 24px;
    margin-top: 80px;
`;

export const SearchWrapper = styled.div`
  @media only screen and (max-width: 540px) {
    width: 100%;
  }
    
`;



