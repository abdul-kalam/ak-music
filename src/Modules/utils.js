export const getLocalStorageValue = (key) => {
  const value = localStorage.getItem(key);
  if (value) {
    return JSON.parse(value);
  }
  return '';
};

export const setLocalStorageValue = (key, value) => {
  value = JSON.stringify(value);
  localStorage.setItem(key, value);
};
