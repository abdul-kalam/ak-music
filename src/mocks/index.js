export const mockTopAlbum = {
  feed: {
    author: {
      name: {
        label: 'iTunes Store',
      },
      uri: {
        label: 'http://www.apple.com/itunes/',
      },
    },
    entry: [
      {
        isFavorite: true,
        'im:name': {
          label: 'Welcome 2 America',
        },
        'im:image': [
          {
            label:
              'https://is4-ssl.mzstatic.com/image/thumb/Music114/v4/55/46/19/55461993-6fd7-d703-9b39-03043f585203/886449187249.jpg/55x55bb.png',
            attributes: {
              height: '55',
            },
          },
          {
            label:
              'https://is3-ssl.mzstatic.com/image/thumb/Music114/v4/55/46/19/55461993-6fd7-d703-9b39-03043f585203/886449187249.jpg/60x60bb.png',
            attributes: {
              height: '60',
            },
          },
          {
            label:
              'https://is1-ssl.mzstatic.com/image/thumb/Music114/v4/55/46/19/55461993-6fd7-d703-9b39-03043f585203/886449187249.jpg/170x170bb.png',
            attributes: {
              height: '170',
            },
          },
        ],
        'im:itemCount': {
          label: '12',
        },
        'im:price': {
          label: '$9.99',
          attributes: {
            amount: '9.99',
            currency: 'USD',
          },
        },
        'im:contentType': {
          'im:contentType': {
            attributes: {
              term: 'Album',
              label: 'Album',
            },
          },
          attributes: {
            term: 'Music',
            label: 'Music',
          },
        },
        rights: {
          label:
            '℗ 2021 NPG Records, Inc. under exclusive license to Legacy Recordings',
        },
        title: {
          label: 'Welcome 2 America - Prince',
        },
        link: {
          attributes: {
            rel: 'alternate',
            type: 'text/html',
            href: 'https://music.apple.com/us/album/welcome-2-america/1561643530?uo=2',
          },
        },
        id: {
          label:
            'https://music.apple.com/us/album/welcome-2-america/1561643530?uo=2',
          attributes: {
            'im:id': '1561643530',
          },
        },
        'im:artist': {
          label: 'Prince',
          attributes: {
            href: 'https://music.apple.com/us/artist/prince/155814?uo=2',
          },
        },
        category: {
          attributes: {
            'im:id': '15',
            term: 'R&B/Soul',
            scheme: 'https://music.apple.com/us/genre/music-r-b-soul/id15?uo=2',
            label: 'R&B/Soul',
          },
        },
        'im:releaseDate': {
          label: '2021-07-30T00:00:00-07:00',
          attributes: {
            label: 'July 30, 2021',
          },
        },
      },
      {
        isFavorite: true,
        'im:name': {
          label: 'Into The Mystery',
        },
        'im:image': [
          {
            label:
              'https://is2-ssl.mzstatic.com/image/thumb/Music125/v4/f5/48/5f/f5485f4c-458e-425e-fa60-e3ce2871cb00/075679788856.jpg/55x55bb.png',
            attributes: {
              height: '55',
            },
          },
          {
            label:
              'https://is4-ssl.mzstatic.com/image/thumb/Music125/v4/f5/48/5f/f5485f4c-458e-425e-fa60-e3ce2871cb00/075679788856.jpg/60x60bb.png',
            attributes: {
              height: '60',
            },
          },
          {
            label:
              'https://is2-ssl.mzstatic.com/image/thumb/Music125/v4/f5/48/5f/f5485f4c-458e-425e-fa60-e3ce2871cb00/075679788856.jpg/170x170bb.png',
            attributes: {
              height: '170',
            },
          },
        ],
        'im:itemCount': {
          label: '12',
        },
        'im:price': {
          label: '$7.99',
          attributes: {
            amount: '7.99',
            currency: 'USD',
          },
        },
        'im:contentType': {
          'im:contentType': {
            attributes: {
              term: 'Album',
              label: 'Album',
            },
          },
          attributes: {
            term: 'Music',
            label: 'Music',
          },
        },
        rights: {
          label: '℗ 2021 Elektra Records LLC',
        },
        title: {
          label: 'Into The Mystery - NEEDTOBREATHE',
        },
        link: {
          attributes: {
            rel: 'alternate',
            type: 'text/html',
            href: 'https://music.apple.com/us/album/into-the-mystery/1566670096?uo=2',
          },
        },
        id: {
          label:
            'https://music.apple.com/us/album/into-the-mystery/1566670096?uo=2',
          attributes: {
            'im:id': '1566670096',
          },
        },
        'im:artist': {
          label: 'NEEDTOBREATHE',
          attributes: {
            href: 'https://music.apple.com/us/artist/needtobreathe/113247214?uo=2',
          },
        },
        category: {
          attributes: {
            'im:id': '21',
            term: 'Rock',
            scheme: 'https://music.apple.com/us/genre/music-rock/id21?uo=2',
            label: 'Rock',
          },
        },
        'im:releaseDate': {
          label: '2021-07-30T00:00:00-07:00',
          attributes: {
            label: 'July 30, 2021',
          },
        },
      },
      {
        'im:name': {
          label: 'The House Is Burning',
        },
        'im:image': [
          {
            label:
              'https://is2-ssl.mzstatic.com/image/thumb/Music115/v4/3f/28/c8/3f28c8e9-c28a-1087-8ccb-654c35bdfbb8/093624878070.jpg/55x55bb.png',
            attributes: {
              height: '55',
            },
          },
          {
            label:
              'https://is4-ssl.mzstatic.com/image/thumb/Music115/v4/3f/28/c8/3f28c8e9-c28a-1087-8ccb-654c35bdfbb8/093624878070.jpg/60x60bb.png',
            attributes: {
              height: '60',
            },
          },
          {
            label:
              'https://is3-ssl.mzstatic.com/image/thumb/Music115/v4/3f/28/c8/3f28c8e9-c28a-1087-8ccb-654c35bdfbb8/093624878070.jpg/170x170bb.png',
            attributes: {
              height: '170',
            },
          },
        ],
        'im:itemCount': {
          label: '16',
        },
        'im:price': {
          label: '$9.99',
          attributes: {
            amount: '9.99',
            currency: 'USD',
          },
        },
        'im:contentType': {
          'im:contentType': {
            attributes: {
              term: 'Album',
              label: 'Album',
            },
          },
          attributes: {
            term: 'Music',
            label: 'Music',
          },
        },
        rights: {
          label:
            '℗ 2021 Top Dawg Entertainment, under exclusive license to Warner Records Inc.',
        },
        title: {
          label: 'The House Is Burning - Isaiah Rashad',
        },
        link: {
          attributes: {
            rel: 'alternate',
            type: 'text/html',
            href: 'https://music.apple.com/us/album/the-house-is-burning/1577158260?uo=2',
          },
        },
        id: {
          label:
            'https://music.apple.com/us/album/the-house-is-burning/1577158260?uo=2',
          attributes: {
            'im:id': '1577158260',
          },
        },
        'im:artist': {
          label: 'Isaiah Rashad',
          attributes: {
            href: 'https://music.apple.com/us/artist/isaiah-rashad/605391263?uo=2',
          },
        },
        category: {
          attributes: {
            'im:id': '18',
            term: 'Hip Hop/Rap',
            scheme:
              'https://music.apple.com/us/genre/music-hip-hop-rap/id18?uo=2',
            label: 'Hip-Hop/Rap',
          },
        },
        'im:releaseDate': {
          label: '2021-07-30T00:00:00-07:00',
          attributes: {
            label: 'July 30, 2021',
          },
        },
      },
    ],
    updated: {
      label: '2021-08-04T13:45:32-07:00',
    },
    rights: {
      label: 'Copyright 2008 Apple Inc.',
    },
    title: {
      label: 'iTunes Store: Top Albums',
    },
    icon: {
      label: 'http://itunes.apple.com/favicon.ico',
    },
    link: [
      {
        attributes: {
          rel: 'alternate',
          type: 'text/html',
          href: 'https://itunes.apple.com/WebObjects/MZStore.woa/wa/viewTop?cc=us&id=1&popId=11',
        },
      },
      {
        attributes: {
          rel: 'self',
          href: 'https://mzstoreservices-int-st.itunes.apple.com/us/rss/topalbums/limit=100/json',
        },
      },
    ],
    id: {
      label:
        'https://mzstoreservices-int-st.itunes.apple.com/us/rss/topalbums/limit=100/json',
    },
  },
  titleListMock: [
    {
      id: '1566670096',
      name: 'Welcome to America',
      image: 'www.test.com/img.jpg',
    },
    {
      id: '1566670094',
      name: 'Welcome to India',
      image: 'www.test.com/img1.jpg',
    },
  ],
};

export const mockAlbumSongs = {
  resultCount: 4,
  results: [
    {
      wrapperType: 'collection',
      collectionType: 'Album',
      artistId: 4779205,
      collectionId: 1570554138,
      amgArtistId: 825908,
      artistName: 'Chris Young',
      collectionName: 'Famous Friends',
      collectionCensoredName: 'Famous Friends',
      artistViewUrl:
        'https://music.apple.com/us/artist/chris-young/4779205?uo=4',
      collectionViewUrl:
        'https://music.apple.com/us/album/famous-friends/1570554138?uo=4',
      artworkUrl60:
        'https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/d1/6e/a9/d16ea9a6-c2a4-90aa-99e7-a981ee8cb689/source/60x60bb.jpg',
      artworkUrl100:
        'https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/d1/6e/a9/d16ea9a6-c2a4-90aa-99e7-a981ee8cb689/source/100x100bb.jpg',
      collectionPrice: 11.99,
      collectionExplicitness: 'notExplicit',
      trackCount: 14,
      copyright: '℗ 2021 Sony Music Entertainment. All rights reserved.',
      country: 'USA',
      currency: 'USD',
      releaseDate: '2021-08-06T07:00:00Z',
      primaryGenreName: 'Country',
    },
    {
      wrapperType: 'track',
      kind: 'song',
      artistId: 4779205,
      collectionId: 1570554138,
      trackId: 1570554142,
      artistName: 'Chris Young',
      collectionName: 'Famous Friends',
      trackName: 'Raised on Country',
      collectionCensoredName: 'Famous Friends',
      trackCensoredName: 'Raised on Country',
      artistViewUrl:
        'https://music.apple.com/us/artist/chris-young/4779205?uo=4',
      collectionViewUrl:
        'https://music.apple.com/us/album/raised-on-country/1570554138?i=1570554142&uo=4',
      trackViewUrl:
        'https://music.apple.com/us/album/raised-on-country/1570554138?i=1570554142&uo=4',
      previewUrl:
        'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview115/v4/10/c0/3f/10c03f67-bb34-ff6b-403b-4c6070fb56ad/mzaf_15785209836383516861.plus.aac.p.m4a',
      artworkUrl30:
        'https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/d1/6e/a9/d16ea9a6-c2a4-90aa-99e7-a981ee8cb689/source/30x30bb.jpg',
      artworkUrl60:
        'https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/d1/6e/a9/d16ea9a6-c2a4-90aa-99e7-a981ee8cb689/source/60x60bb.jpg',
      artworkUrl100:
        'https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/d1/6e/a9/d16ea9a6-c2a4-90aa-99e7-a981ee8cb689/source/100x100bb.jpg',
      collectionPrice: 11.99,
      trackPrice: 1.29,
      releaseDate: '2019-01-04T08:00:00Z',
      collectionExplicitness: 'notExplicit',
      trackExplicitness: 'notExplicit',
      discCount: 1,
      discNumber: 1,
      trackCount: 14,
      trackNumber: 1,
      trackTimeMillis: 176520,
      country: 'USA',
      currency: 'USD',
      primaryGenreName: 'Country',
      isStreamable: true,
    },
    {
      wrapperType: 'track',
      kind: 'song',
      artistId: 4779205,
      collectionId: 1570554138,
      trackId: 1570554143,
      artistName: 'Chris Young & Kane Brown',
      collectionName: 'Famous Friends',
      trackName: 'Famous Friends',
      collectionCensoredName: 'Famous Friends',
      trackCensoredName: 'Famous Friends',
      collectionArtistName: 'Chris Young',
      artistViewUrl:
        'https://music.apple.com/us/artist/chris-young/4779205?uo=4',
      collectionViewUrl:
        'https://music.apple.com/us/album/famous-friends/1570554138?i=1570554143&uo=4',
      trackViewUrl:
        'https://music.apple.com/us/album/famous-friends/1570554138?i=1570554143&uo=4',
      previewUrl:
        'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview125/v4/42/7a/5f/427a5f4b-5e18-fdb6-3ba8-56a2d06483b0/mzaf_2809262009335345042.plus.aac.p.m4a',
      artworkUrl30:
        'https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/d1/6e/a9/d16ea9a6-c2a4-90aa-99e7-a981ee8cb689/source/30x30bb.jpg',
      artworkUrl60:
        'https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/d1/6e/a9/d16ea9a6-c2a4-90aa-99e7-a981ee8cb689/source/60x60bb.jpg',
      artworkUrl100:
        'https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/d1/6e/a9/d16ea9a6-c2a4-90aa-99e7-a981ee8cb689/source/100x100bb.jpg',
      collectionPrice: 11.99,
      trackPrice: 1.29,
      releaseDate: '2020-11-20T08:00:00Z',
      collectionExplicitness: 'notExplicit',
      trackExplicitness: 'notExplicit',
      discCount: 1,
      discNumber: 1,
      trackCount: 14,
      trackNumber: 2,
      trackTimeMillis: 166507,
      country: 'USA',
      currency: 'USD',
      primaryGenreName: 'Country',
      isStreamable: true,
    },
    {
      wrapperType: 'track',
      kind: 'song',
      artistId: 4779205,
      collectionId: 1570554138,
      trackId: 1570554144,
      artistName: 'Chris Young & Lauren Alaina',
      collectionName: 'Famous Friends',
      trackName: "Town Ain't Big Enough",
      collectionCensoredName: 'Famous Friends',
      trackCensoredName: "Town Ain't Big Enough",
      collectionArtistName: 'Chris Young',
      artistViewUrl:
        'https://music.apple.com/us/artist/chris-young/4779205?uo=4',
      collectionViewUrl:
        'https://music.apple.com/us/album/town-aint-big-enough/1570554138?i=1570554144&uo=4',
      trackViewUrl:
        'https://music.apple.com/us/album/town-aint-big-enough/1570554138?i=1570554144&uo=4',
      previewUrl:
        'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview115/v4/d3/0d/69/d30d693f-570d-0bc0-3c70-7c7a272baec3/mzaf_7645675846915846385.plus.aac.p.m4a',
      artworkUrl30:
        'https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/d1/6e/a9/d16ea9a6-c2a4-90aa-99e7-a981ee8cb689/source/30x30bb.jpg',
      artworkUrl60:
        'https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/d1/6e/a9/d16ea9a6-c2a4-90aa-99e7-a981ee8cb689/source/60x60bb.jpg',
      artworkUrl100:
        'https://is1-ssl.mzstatic.com/image/thumb/Music115/v4/d1/6e/a9/d16ea9a6-c2a4-90aa-99e7-a981ee8cb689/source/100x100bb.jpg',
      collectionPrice: 11.99,
      trackPrice: 1.29,
      releaseDate: '2019-07-12T07:00:00Z',
      collectionExplicitness: 'notExplicit',
      trackExplicitness: 'notExplicit',
      discCount: 1,
      discNumber: 1,
      trackCount: 14,
      trackNumber: 3,
      trackTimeMillis: 171800,
      country: 'USA',
      currency: 'USD',
      primaryGenreName: 'Country',
      isStreamable: true,
    },
  ],
};
