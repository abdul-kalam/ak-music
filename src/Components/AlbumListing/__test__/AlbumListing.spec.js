import React from 'react';
import renderer from 'react-test-renderer';
import { AlbumListing } from './../AlbumListing';
import { mockTopAlbum } from '../../../mocks';
import { render, screen } from '../../../Modules/test-utils';
import * as reactRedux from 'react-redux'

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn().mockImplementationOnce(() => {
    return jest.fn();
  }),
}));

jest.mock('./../../../Components/Album/Album', () => ({
  Album: () => {
    return <div>Album List rendered</div>;
  },
}));

jest.mock('../../../Components/Autocomplete/Autocomplete', () => ({
  Autocomplete: () => {
    return (
      <select>
        <option value="all">All</option>
        <option value="10">Rock</option>
        <option value="12">Zazz</option>
      </select>
    );
  },
}));

jest.mock('../../../Components/Filter/filter', () => ({
  Filter: () => {
    return <div>Filter rendered</div>;
  },
}));

describe('Album listing page', () => {
  const useSelectorMock = jest.spyOn(reactRedux, 'useSelector')
  const useDispatchMock = jest.spyOn(reactRedux, 'useDispatch')
  beforeEach(() => {
    useSelectorMock.mockClear()
    useDispatchMock.mockClear()
  })
  test('component should render with list of albums', () => {
    useSelectorMock.mockReturnValue(mockTopAlbum.feed.entry)
    const { asFragment } = render(<AlbumListing />);

    expect(screen.getAllByText('Album List rendered')).toHaveLength(3);
    expect(asFragment()).toMatchSnapshot();
  });

  test('component should render with message info if albums not available', () => {
    useSelectorMock.mockReturnValue([])
    const { asFragment } = render(<AlbumListing />);

    expect(screen.getByText('Albums not available for selected criteria. Try something else!')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });


});
