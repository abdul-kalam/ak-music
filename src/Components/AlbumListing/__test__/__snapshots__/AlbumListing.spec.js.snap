// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Album listing page component should render with list of albums 1`] = `
<DocumentFragment>
  <section
    class="sc-dlnjwi FNGXW"
  >
    <div
      class="sc-hKFxyN dZTiSr"
    >
      <select>
        <option
          value="all"
        >
          All
        </option>
        <option
          value="10"
        >
          Rock
        </option>
        <option
          value="12"
        >
          Zazz
        </option>
      </select>
    </div>
    <div>
      Filter rendered
    </div>
    <section
      class="sc-gtsrHT hMxucw"
    >
      <ul
        class="sc-bdnxRM gytOic"
      >
        <div>
          Album List rendered
        </div>
        <div>
          Album List rendered
        </div>
        <div>
          Album List rendered
        </div>
      </ul>
    </section>
  </section>
</DocumentFragment>
`;

exports[`Album listing page component should render with message info if albums not available 1`] = `
<DocumentFragment>
  <section
    class="sc-dlnjwi FNGXW"
  >
    <div
      class="sc-hKFxyN dZTiSr"
    >
      <select>
        <option
          value="all"
        >
          All
        </option>
        <option
          value="10"
        >
          Rock
        </option>
        <option
          value="12"
        >
          Zazz
        </option>
      </select>
    </div>
    <div>
      Filter rendered
    </div>
    <section
      class="sc-gtsrHT hMxucw"
    >
      <div
        class="sc-eCApnc cfSDGH"
        type="info"
      >
        <p
          class="sc-jSFjdj jYHEXi"
        >
          Albums not available for selected criteria. Try something else!
        </p>
      </div>
    </section>
  </section>
</DocumentFragment>
`;
